pkg/kube/services/services.go

```diff
@@ -92,7 +95,42 @@ func GetServiceURLFromMap(services map[string]*v1.Service, name string) string {
 	return GetServiceURL(services[name])
 }
 
-func FindServiceURL(client kubernetes.Interface, namespace string, name string) (string, error) {
+func getIstioVirtualService(dynamicClient dynamic.Interface, namespace, name string) (*unstructured.Unstructured, error) {
+	dynamicClient, err := kube.LazyCreateDynamicClient(dynamicClient)
+	if err != nil {
+		return nil, err
+	}
+	//  Create a GVR which represents an Istio Virtual Service.
+	virtualServiceGVR := schema.GroupVersionResource{
+		Group:    "networking.istio.io",
+		Version:  "v1alpha3",
+		Resource: "virtualservices",
+	}
+	virtualService, err := dynamicClient.Resource(virtualServiceGVR).Namespace(namespace).Get(context.TODO(), name, meta_v1.GetOptions{})
+	return virtualService, err
+}
+
+// getUrlFromVirtualService finds url from virtual services istio
+func getUrlFromVirtualService(virtualService *unstructured.Unstructured) (string, error) {
+	vs := virtualService.Object
+	if spec, ok := vs["spec"].(map[string]interface{}); ok {
+		if hosts, ok := spec["hosts"].([]interface{}); ok {
+			return fmt.Sprintf("%v", hosts[0]), nil
+		}
+	}
+	return "", errors.New("No url found in the virtual service")
+}
+
+// FindUrlFromVsIstio finds the host from istio virtual service
+func FindUrlFromVsIstio(dynamicClient dynamic.Interface, namespace, name string) (string, error) {
+	virtualService, err := getIstioVirtualService(dynamicClient, namespace, name)
+	if err != nil {
+		return "", err
+	}
+	return getUrlFromVirtualService(virtualService)
+}
+
+func FindServiceURL(client kubernetes.Interface, namespace string, name string, dynamicClient dynamic.Interface) (string, error) {
```

Comment:
@msvticket brought a good point, can we keep the function signature same, and add a completely new function (you also suggested this).

Reply:
For that we will have to create dynamicClient inside getIstioVirtualService and that would happen only if I didn't get service url from ingress.
Should we do that?

Reply:
or in jx-preview if we don;t get url from FindServiceURL then we can call FindUrlFromVsIstio and can pass dynamicClient from there.But i personally like the above approach : )

Reply:
>or in jx-preview if we don;t get url from FindServiceURL then we can call FindUrlFromVsIstio and can pass dynamicClient from there.But i personally like the above approach : )

Let's do this! You already have the tests for this.

Reply:
The method is called from other applications than jx-preview. I would prefer something like:

```
 func FindServiceURL(client kubernetes.Interface, namespace string, name string) (string, error) {
  return FindServiceURLWithDynamicClient(client, namespace, name, nil)
}

// The new function
 func FindServiceURLWithDynamicClient(client kubernetes.Interface, namespace string, name string, dynamicClient dynamic.Interface) (string, error) {
...
}

```

Then you can use FindServiceURLWithDynamicClient in the tests and the code using FindServiceURL will get the new functionality without code change.

Reply:
Done